src/eval/context.rs

```diff
@@ -0,0 +1,98 @@
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+
+use super::{Env, eval};
+use ::{Value, AresResult, AresError, parse, stdlib, Environment};
+
+pub struct Context<T> {
+    env: Env,
+    _state: PhantomData<T>,
+}
+
+pub struct LoadedContext<'a, 'b, T: 'a + 'b> {
+    ctx: &'b mut Context<T>,
+    state: &'a mut T,
+}
+
+impl <T> Context<T> {
+    pub fn new() -> Context<T> {
+        let env = Rc::new(RefCell::new(Environment::new()));
+        stdlib::load_all(&env);
+        Context {
+            env: env,
+            _state: PhantomData,
+        }
+    }
+
+    pub fn new_empty() -> Context<T> {
+        Context {
+            env: Rc::new(RefCell::new(Environment::new())),
+            _state: PhantomData,
+        }
+    }
+
+    pub fn load<'a, 'b>(&'b mut self, state: &'a mut T) -> LoadedContext<'a, 'b, T> {
+        LoadedContext {
+            ctx: self,
+            state: state
+        }
+    }
+
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+}
+
+impl <'a, 'b,  T> LoadedContext<'a, 'b, T> {
+    pub fn eval(&mut self, program: &Value) -> AresResult<Value> {
+        eval(program, self.env())
+    }
+
+    pub fn eval_str(&mut self, program: &str) -> AresResult<Value> {
+        let trees = try!(parse(program));
+        let mut last = None;
+        for tree in trees {
+            last = Some(try!(eval(&tree, self.env())))
+        }
+        match last {
+            Some(v) => Ok(v),
+            None => Err(AresError::NoProgram)
+        }
+    }
+
+    pub fn call(&mut self, func: Value, mut args: Vec<Value>) -> AresResult<Value> {
+        for arg in args.iter_mut() {
+            if let a@&mut Value::List(_) = arg {
+                // TODO: can you get rid of this clone?
+                *a = Value::new_list(vec![Value::new_ident("quote"), a.clone()]);
+            }
+        }
+
+        let built = Value::new_list(vec![func, Value::new_list(args)]);
+        self.eval(&built)
```

Comment:
I suspect the `clone()` can be got rid of, and this whole method made less roundabout, if the part of `eval` that does applies the function to already-evaluated arguments is abstracted out and called directly here. We already know (or at least assume!) that `func` is a procedure, so we shouldn't need to create a list in `built` just to match on it in `eval`, and could also avoid quoting to avoid double-evaluation of lists in the arguments.